crates/noirc_evaluator/src/ssa/acir_gen/acir_mem.rs

```diff
@@ -1,22 +1,158 @@
-use crate::ssa::{
-    acir_gen::InternalVar,
-    context::SsaContext,
-    mem::{ArrayId, MemArray},
+use crate::{
+    ssa::{
+        acir_gen::InternalVar,
+        context::SsaContext,
+        mem::{ArrayId, MemArray},
+    },
+    Evaluator,
 };
-use acvm::acir::native_types::Witness;
+use acvm::{
+    acir::{
+        circuit::{directives::Directive, opcodes::Opcode as AcirOpcode},
+        native_types::{Expression, Witness},
+    },
+    FieldElement,
+};
+
 use iter_extended::vecmap;
 use std::collections::BTreeMap;
 
+use super::{
+    constraints::{self, mul_with_witness, subtract},
+    expression_from_witness,
+    operations::{self},
+};
+
+#[derive(Clone, Debug)]
+struct MemItem {
+    op: Expression,
+    value: Expression,
+    index: Expression,
+}
+
 #[derive(Default)]
 pub struct ArrayHeap {
     // maps memory address to InternalVar
     memory_map: BTreeMap<u32, InternalVar>,
+    trace: Vec<MemItem>,
```

Comment:
I think it is fine as the map is private and only accessible through a few functions, and also the index is often converted to field elements/expression so this would add a lot of conversions between u32, memaddress and expressions.

Reply:
This wouldn't add any conversions, the suggestion I gave is a type alias rather than a newtype

Reply:
ok, done.